File: OOP C++/lab2/If.cpp
#include "If.h"
#include "FactoryComplexFuncInitializer.h"
#include "Utility.h"
#include <memory>

void If::Execute(Operands& operands, Tokens& tokens, std::string& output, Reader& reader) {
    auto pFactory = Factory<Command, std::string, Command *(*)()>::getInstance();
    Tokens tempTokens(tokens);

    int operand = operands.Top();
    if (operand == 0) {
        while (!tokens.IsEmpty()) {
            std::string token = tokens.GetAndPop();
            if(pFactory->isComplexFuncRegist3red(token)) {
                auto pCommand = std::unique_ptr<Command>(pFactory->createProductByName(token));
                pCommand->Check(tokens, reader);
            } else if (token == "else") {
                while (!tokens.IsEmpty()) {
                    token = tokens.GetAndPop();
                    if (isNumber(token)) {
                        operands.Push(std::stoi(token));
                    }
                    else if(pFactory->isRegist3red(token)) {
                        auto pCommand = std::unique_ptr<Command>(pFactory->createProductByName(token));
                        pCommand->Execute(operands, tokens, output, reader);
                    } else if (token == "else") {
                        throw std::runtime_error("Incorrect number of \"else\"");
                    } else if (token == "then") {
                        CheckSemicolon(tokens);
                        return;
                    } else {
                        token += " ?";
                        throw std::runtime_error(token);
                    }
                }
            } else if (token == "then") {
                CheckSemicolon(tokens);
                return;
            } else if(!pFactory->isRegist3red(token) && !isNumber(token)) {
                token += " ?";
                throw std::runtime_error(token);
            }
        }
    }
    else {
        while (!tokens.IsEmpty()) {
            std::string token = tokens.GetAndPop();
            if (isNumber(token)) {
                operands.Push(std::stoi(token));
            } else if(pFactory->isRegist3red(token)) {
                auto pCommand = std::unique_ptr<Command>(pFactory->createProductByName(token));
                pCommand->Execute(operands, tokens, output, reader);
            } else if(token == "else") {
                while(!tokens.IsEmpty()) {
                    token = tokens.GetAndPop();
                    if(pFactory->isComplexFuncRegist3red(token)) {
                        auto pCommand = std::unique_ptr<Command>(pFactory->createProductByName(token));
                        pCommand->Check(tokens, reader);
                    } else if(token == "else") {
                        throw std::runtime_error("Incorrect number of \"else\"");
                    } else if (token == "then") {
                        CheckSemicolon(tokens);
                        return;
                    } else if(!pFactory->isRegist3red(token) && !isNumber(token)) {
                        token += " ?";
                        throw std::runtime_error(token);
                    }
                }
            } else if(token == "then") {
                CheckSemicolon(tokens);
                return;
            } else {
                token += " ?";
                throw std::runtime_error(token);
            }
        }
    }
    throw std::runtime_error("There is no then at the end of condition");
}

void If::Check(Tokens& tokens, Reader& reader) {
    auto pFactory = Factory<Command, std::string, Command *(*)()>::getInstance();
    while (!tokens.IsEmpty()) {
        std::string token = tokens.GetAndPop();
        if(pFactory->isComplexFuncRegist3red(token)) {
            auto pCommand = std::unique_ptr<Command>(pFactory->createProductByName(token));
            pCommand->Check(tokens, reader);
        }
        else if (token == "else") {
            while (!tokens.IsEmpty()) {
                token = tokens.GetAndPop();
                if(pFactory->isComplexFuncRegist3red(token)) {
                    auto pCommand = std::unique_ptr<Command>(pFactory->createProductByName(token));
                    pCommand->Check(tokens, reader);
                } else if (token == "else") {
                    throw std::runtime_error("Incorrect number of \"else\"");
                } else if (token == "then") {
                    CheckSemicolon(tokens);
                    return;
                } else if (!isNumber(token) && !pFactory->isRegist3red(token)){
                    token += " ?";
                    throw std::runtime_error(token);
                }
            }
        } else if (token == "then") {
            CheckSemicolon(tokens);
            return;
        } else if(!isNumber(token) && !pFactory->isRegist3red(token)) {
            token += " ?";
            throw std::runtime_error(token);
        }
    }
    throw std::runtime_error("Incorrect form of condition");
}

namespace {
    FactoryComplexFuncInitializer<If> Registration("if");
}

File: OOP Java/lab1/src/main/java/org/example/commands/If.java
package org.example.commands;

import org.example.utility.Operands;

import org.example.factory.Factory;
import org.example.utility.Utility;

import java.util.StringTokenizer;

public class If implements Command{
    @Override
    public void Execute(StringTokenizer tokens, Operands operands) {
        Factory factory = Factory.getInstance();
        Utility utility = new Utility();
        int num = operands.Top();
        if(num == 0) {
            while(tokens.hasMoreTokens()) {
                String token = tokens.nextToken();
                if(factory.isRegistered(token)) {
                    Command command = factory.createCommand(token);
                    command.Check(tokens, operands);
                } else if(token.equals("else")) {
                    while(tokens.hasMoreTokens()) {
                        token = tokens.nextToken();
                        if(utility.isNumber(token)) {
                            operands.Push(Integer.parseInt(token));
                        } else if(factory.isRegistered(token)) {
                            Command command = factory.createCommand(token);
                            command.Execute(tokens, operands);
                        } else if(token.equals("else")) {
                            throw new RuntimeException("Incorrect number of \"else\"");
                        } else if(token.equals("then")) {
                            utility.CheckSemiColumn(tokens);
                            return;
                        } else {
                            throw new RuntimeException(token + " ?");
                        }
                    }
                } else if(token.equals("then")) {
                    utility.CheckSemiColumn(tokens);
                    return;
                } else if(!utility.isNumber(token)) {
                    throw new RuntimeException(token + " ?");
                }
            }
        }
        else {
            while(tokens.hasMoreTokens()) {
                String token = tokens.nextToken();
                if(utility.isNumber(token)) {
                    operands.Push(Integer.parseInt(token));
                } else if(factory.isRegistered(token)) {
                    Command command = factory.createCommand(token);
                    command.Execute(tokens, operands);
                } else if(token.equals("else")) {
                    while(tokens.hasMoreTokens()) {
                        token = tokens.nextToken();
                        if(factory.isRegistered(token)) {
                            Command command = factory.createCommand(token);
                            command.Check(tokens, operands);
                        } else if(token.equals("else")) {
                            throw new RuntimeException("Incorrect number of \"else\"");
                        } else if(token.equals("then")) {
                            utility.CheckSemiColumn(tokens);
                            return;
                        } else if(!utility.isNumber(token)) {
                            throw new RuntimeException(token + " ?");
                        }
                    }
                } else if(token.equals("then")) {
                    utility.CheckSemiColumn(tokens);
                    return;
                } else {
                    throw new RuntimeException(token + " ?");
                }
            }
        }
    }

    @Override
    public void Check(StringTokenizer tokens, Operands operands) {
        Factory factory = Factory.getInstance();
        Utility utility = new Utility();
        int num = operands.Top();
        while(tokens.hasMoreTokens()) {
            String token = tokens.nextToken();
            if(factory.isRegistered(token)) {
                Command command = factory.createCommand(token);
                command.Check(tokens, operands);
            } else if(token.equals("else")) {
                while(tokens.hasMoreTokens()) {
                    token = tokens.nextToken();
                    if(factory.isRegistered(token)) {
                        Command command = factory.createCommand(token);
                        command.Check(tokens, operands);
                    } else if(token.equals("else")) {
                        throw new RuntimeException("Incorrect number of \"else\"");
                    } else if(token.equals("then")) {
                        utility.CheckSemiColumn(tokens);
                        return;
                    } else if(!utility.isNumber(token)) {
                        throw new RuntimeException(token + " ?");
                    }
                }
            } else if(token.equals("then")) {
                utility.CheckSemiColumn(tokens);
                return;
            } else if(!utility.isNumber(token)) {
                throw new RuntimeException(token + " ?");
            }
        }
        throw new RuntimeException("Incorrect form of condition");
    }

}
